framework/app.mk

```diff
@@ -79,14 +79,16 @@ $(eval $(call unity_dir_rule, $(unity_src_dir)))
 # The idea here is that if all they have is src then it's a big jumble of stuff
 # that won't benefit from unity building
 # Also, exclude the base directory by default because it's another big jumble
-# of unrelated stuff
-non_unity_dirs := %.libs %/src %src/base $(app_non_unity_dirs)
+# of unrelated stuff.
+# Note: If a top-level directory is excluded from being built with unity, we
+# ignore contents of all subdirectories under that folder.
+non_unity_dirs := %.libs %/src $(app_non_unity_dirs)
```

Comment:
How about utilize a new variable like `app_level1_unity_dirs`? It will be more complicated than this simple change, but it can reserve the current behavior.

Reply:
There’s already an application level variable you can set.

Reply:
Are you saying that non-unity variable? This new variable is to specify what directories should be considered in unity build but lump all files under it including its sub-folders. Maybe I misunderstood something?

Reply:
I was suggesting something like this:
```
diff --git a/framework/app.mk b/framework/app.mk
index 4c7672576c..6251e6f014 100644
--- a/framework/app.mk
+++ b/framework/app.mk
@@ -80,7 +80,7 @@ $(eval $(call unity_dir_rule, $(unity_src_dir)))
 # that won't benefit from unity building
 # Also, exclude the base directory by default because it's another big jumble
 # of unrelated stuff
-non_unity_dirs := %.libs %/src %src/base $(app_non_unity_dirs)
+non_unity_dirs := %.libs %/src $(app_non_unity_dirs)
 
 # Find all of the individual subdirectories
 # We will create a Unity file for each individual subdirectory
@@ -88,8 +88,11 @@ non_unity_dirs := %.libs %/src %src/base $(app_non_unity_dirs)
 # and will benefit from a Unity build
 srcsubdirs := $(shell find $(APPLICATION_DIR)/src -type d -not -path '*/.libs*')
 
+level1_srcsubdirs := $(shell find $(app_level1_unity_dirs) -type d -not -path '*/.libs*')
+
 # Filter out the paths we don't want to Unity build
 unity_srcsubdirs := $(filter-out $(non_unity_dirs), $(srcsubdirs))
+unity_srcsubdirs := $(filter-out $(level1_srcsubdirs), $(unity_srcsubdirs))
 non_unity_srcsubdirs := $(filter $(non_unity_dirs), $(srcsubdirs))
 
 # This is a biggie
@@ -98,8 +101,11 @@ non_unity_srcsubdirs := $(filter $(non_unity_dirs), $(srcsubdirs))
 # full hierarchy of the path underneath src
 $(foreach srcsubdir,$(unity_srcsubdirs),$(eval $(call unity_file_rule,$(call unity_unique_name,$(unity_src_dir),$(APPLICATION_DIR),$(srcsubdir)),$(shell find $(srcsubdir) -maxdepth 1 \( -type f -o -type l \) -regex "[^\#~]*\.C"),$(srcsubdir),$(unity_src_dir))))
 
+$(foreach srcsubdir,$(app_level1_unity_dirs),$(eval $(call unity_file_rule,$(call unity_unique_name,$(unity_src_dir),$(APPLICATION_DIR),$(srcsubdir)),$(shell find $(srcsubdir) \( -type f -o -type l \) -regex "[^\#~]*\.C"),$(srcsubdir),$(unity_src_dir))))
+
 # This creates the whole list of Unity source files so we can use it as a dependency
 app_unity_srcfiles := $(foreach srcsubdir,$(unity_srcsubdirs),$(call unity_unique_name,$(unity_src_dir),$(APPLICATION_DIR),$(srcsubdir)))
+app_unity_srcfiles += $(foreach srcsubdir,$(app_level1_unity_dirs),$(call unity_unique_name,$(unity_src_dir),$(APPLICATION_DIR),$(srcsubdir)))
 
 # Add to the global list of unity source files
 unity_srcfiles += $(app_unity_srcfiles)
```

Reply:
@YaqiWang - This PR already changes the existing behavior so that we lump all depth = 1 data together across the board, which is exactly what you want for Griffin. It halves the compile time. It's most of the way there! I think the only thing we need to address now are exceptions. For that we have `$app_non_unity_dirs` which should do the right thing if we need to exclude a specific directory in an application. 

IMO - We don't need any other logic at the top level, as @dschwen pointed out, I'll need to exclude the libTorch directories, but hopefully that's about it unless there are other application level conflicts.

Am I going to finish this up? Probably won't get another chance until about the middle of the December! If nobody else gets to it before then, I'll pick it back up!

Reply:
@YaqiWang - This PR already changes the existing behavior so that we lump all depth = 1 data together across the board, which is exactly what you want for Griffin. It halves the compile time. It's most of the way there! I think the only thing we need to address now are exceptions. For that we have $app_non_unity_dirs which should do the right thing if we need to exclude a specific directory in an application.

IMO - We don't need any other logic at the top level, as @dschwen pointed out, I'll need to exclude the libTorch directories, but hopefully that's about it unless there are other application level conflicts.

Am I going to finish this up? Probably won't get another chance until about the middle of the December! If nobody else gets to it before then, I'll pick it back up!